### PR TITLE
Ssl credentials override default roots

### PR DESCRIPTION
**Pull request by jboeuf**
Take 2. Surviving the outage....

Fixes the first part of #4834.

### CODE REVIEW

**Comment by jboeuf** [comment]
@jtattermusch for review as well.

File: include/grpc/grpc_security.h [hunk 1]
```diff
@@ -143,6 +143,15 @@ grpc_channel_credentials *grpc_google_default_credentials_create(void);
...
   "GRPC_DEFAULT_SSL_ROOTS_FILE_PATH"
 
+/* Overrides the default TLS/SSL roots.
+   The roots must be encoded as PEM and NULL-terminated.
+   This function is not thread-safe and must be called at initialization time
+   before any ssl credentials are created to have the desired side effect.
+   It also does not do any checks about the validity of the encoding.
+   If the GRPC_DEFAULT_SSL_ROOTS_FILE_PATH environment is set to a valid path,
+   it will override the roots specified in this function. */
+void grpc_override_ssl_default_roots(const char *roots_pem);
```
Comment (jtattermusch) [inline include/grpc/grpc_security.h]: Nit:
consider calling the arg  "pem_root_certs" as this is the name of corresponding arg in grpc_ssl_credentials_create, they represent the same thing, so using the same name facilitates understanding.

File: src/core/security/security_connector.c [hunk 2]
```diff
@@ -595,23 +603,38 @@ static grpc_security_connector_vtable ssl_channel_vtable = {
...
-  if (default_root_certs_path == NULL) {
-    default_pem_root_certs = gpr_empty_slice();
-  } else {
-    default_pem_root_certs = gpr_load_file(default_root_certs_path, 0, NULL);
+  if (default_root_certs_path != NULL) {
+    result = gpr_load_file(default_root_certs_path, 0, NULL);
     gpr_free(default_root_certs_path);
   }
 
+  /* Try overridden roots path if needed. */
```
Comment (jtattermusch) [inline src/core/security/security_connector.c]: s/path//

**Comment by jtattermusch** [comment]
Besides a minor nit, the API looks good. Also made a quick pass over the code, but a review from second pair of eyes would still be useful, as this is security.

**Comment by jtattermusch** [comment]
Also, run tools/distrib/check_copyright.py --fix, some copyrights need updating.

**Comment by ctiller** [comment]
We just did a round trip on this...

suggest (as a public API):
```c
enum Result {
  OK,
  FAIL_PERMANENTLY, // don't try any fallback options
  FAIL,
};
// fn has the semantics that if it returns ok, 
// *pem_file is the root list to be owned by core and
// freed with gpr_free()
void SetPEMOverride(Result (*fn)(char **pem_file));
```

**Comment by ctiller** [comment]
(I think implementation details can remain as-is for now - just call the callback immediately)

**Comment by jboeuf** [comment]
changed as discussed. PTAL.

**Comment by murgatroid99** [comment]
Can we make `src/core/support/file.h` a public header? In wrapped languages that aren't C#, the callback will likely be "read a roots.pem file, and return the contents"

**Comment by jboeuf** [comment]
@murgatroid99. You should probably not use `gpr_load_file` as it returns a `gpr_slice` and not an allocated `char *` that needs to be freed with `gpr_free`.

**Comment by jboeuf** [comment]
Jenkins is mostly green.

**Pull request merged**